### ISSUE TITLE
C0103: Variable name false positive for single word all caps at module level

### ISSUE THREAD

**Issue by fmigneault** [PyCQA/pylint#11012]
### Bug description

The file shown in the error contains the following at module level:
```python
HTML = lxml_etree.HTML
```
https://github.com/crim-ca/weaver/blob/8ec035906007ede756f2fb91874feeec58953a00/weaver/xml_util.py#L70-L72

It doesn't look like it should raise a "variable name" error, since it is all caps (constant at module level).

Maybe because the assigned value is not a literal? I don't think it should be considered invalid.

### Configuration
https://github.com/crim-ca/weaver/blob/refs/heads/dependabot/pip/pylint-gte-4.0.5-and-lt-5/.pylintrc
```ini


[MASTER]

# A comma-separated list of package or module names from where C extensions may
# be loaded. Extensions are loading into the active Python interpreter and may
# run arbitrary code.
extension-pkg-whitelist=lxml.etree,
                        schema_salad.sourceline,
                        schema_salad.validate

# Add files or directories to the blacklist. They should be base names, not
# paths.
ignore=CVS

# Add files or directories matching the regex patterns to the blacklist. The
# regex matches against base names, not paths.
ignore-patterns=

# Python code to execute, usually for sys.path manipulation such as
# pygtk.require().
#init-hook=

# Use multiple processes to speed up Pylint. Specifying 0 will auto-detect the
# number of processors available to use.
jobs=0

# Control the amount of potential inferred values when inferring a single
# object. This can help the performance when dealing with large functions or
# complex, nested conditions.
limit-inference-results=100

# List of plugins (as comma separated values of python module names) to load,
# usually to register additional checkers.
# https://pylint.pycqa.org/en/latest/technical_reference/extensions.html
load-plugins=pylint.extensions.docparams,
             pylint.extensions.mccabe,
             pylint_per_file_ignores

# https://pylint.pycqa.org/en/latest/technical_reference/extensions.html#design-checker-options
max-complexity = 24
# FIXME: medium/lower-high complexity permitted, should focus toward <20 to facilitate testing and debugging

# https://pylint.pycqa.org/en/latest/technical_reference/extensions.html#parameter-documentation-checker-options
# allow unspecified items, but if the docstring is present, it should follow Sphinx format for documentation generation
# see also 'disabled' W90XX warnings, we ignore only raised 'missing', but detect invalid documentation issues
accept-no-param-doc = yes
accept-no-raise-doc = yes
accept-no-return-doc = yes
accept-no-yields-doc = yes
default-docstring-type = sphinx

# Pickle collected data for later comparisons.
persistent=yes

# Specify a configuration file.
#rcfile=

# When enabled, pylint would attempt to guess common misconfiguration and emit
# user-friendly hints instead of false-positive error messages.
suggestion-mode=yes

# Allow loading of arbitrary C extensions. Extensions are imported into the
# active Python interpreter and may run arbitrary code.
unsafe-load-any-extension=no


[MESSAGES CONTROL]

# Only show warnings with the listed confidence levels. Leave empty to show
# all. Valid levels: HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED.
confidence=

# Disable the message, report, category or checker with the given id(s). You
# can either give multiple identifiers separated by comma (,) or put this
# option multiple times (only on the command line, not in the configuration
# file where it should appear only once). You can also use "--disable=all" to
# disable everything first and then re-enable specific checks. For example, if
# you want to run only the similarities checker, you can use "--disable=all
# --enable=similarities". If you want to run only the classes checker, but have
# no Warning level messages displayed, use "--disable=all --enable=classes
# --disable=W".
disable=C0111,missing-docstring,
        C0206,consider-using-dict-items,
        C0302,too-many-lines,
        C0412,ungrouped-imports,
        C0415,import-outside-toplevel,
        C1801,len-as-condition,
        E0015,unrecognized-option,
        E0401,import-error,
        E5110,django-not-configured,
        R0022,useless-option-value,
        R0201,no-self-use,
        R0205,useless-object-inheritance,
        # R0401 disabled to avoid unnecessary warnings on already handled circular imports
        # See issue : https://github.com/PyCQA/pylint/issues/850
        R0401,cyclic-import,
        R0801,duplicate-code,
        R0901,too-many-ancestors,
        R0902,too-many-instance-attributes,
        R0904,too-many-public-methods,
        R0911,too-many-return-statements,
        R0912,too-many-branches,
        R0913,too-many-arguments,
        R0914,too-many-locals,
        R0915,too-many-statements,
        R0917,too-many-positional-arguments,
        R1705,no-else-return,
        R1710,inconsistent-return-statements,
        R1711,useless-return,
        R1721,unnecessary-comprehension,
        R1725,super-with-arguments,
        W0012,unknown-option-value,
        W0108,unnecessary-lambda,
        W0212,protected-access,
        W0232,no-init,
        W0235,useless-super-delegation,
        W0613,unused-argument,
        W0622,redefined-builtin,
        W0640,cell-var-from-loop,
        W0706,try-except-raise,
        W0707,raise-missing-from,
        W1508,invalid-envvar-default,
        W9005,multiple-constructor-doc,
        W9006,missing-raises-doc,
        W9011,missing-return-doc,
        W9012,missing-return-type-doc,
        W9013,missing-yield-doc,
        W9014,missing-yield-type-doc,
        W9015,missing-param-doc,
        W9016,missing-type-doc

# note: (C0412, ungrouped-imports) is managed via isort tool, ignore false positives

per-file-ignores =
  tests/*:R1729

# Enable the message, report, category or checker with the given id(s). You can
# either give multiple identifier separated by comma (,) or put this option
# multiple time (only on the command line, not in the configuration file where
# it should appear only once). See also the "--disable" option for examples.
enable=c-extension-no-member


[REPORTS]

# Python expression which should return a score less than or equal to 10. You
# have access to the variables 'error', 'warning', 'refactor', and 'convention'
# which contain the number of messages in each category, as well as 'statement'
# which is the total number of statements analyzed. This score is used by the
# global evaluation report (RP0004).
evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)

# Template used to display messages. This is a python new-style format string
# used to format the message information. See doc for all details.
#msg-template=

# Set the output format. Available formats are text, parseable, colorized, json
# and MSVC (visual studio). You can also give a reporter class, e.g.
# mypackage.mymodule.MyReporterClass.
output-format=colorized

# Tells whether to display a full report or only the messages.
reports=no

# Activate the evaluation score.
score=yes


[REFACTORING]

# Maximum number of nested blocks for function / method body
max-nested-blocks=10

# Complete name of functions that never returns. When checking for
# inconsistent-return-statements if a never returning function is called then
# it will be considered as an explicit return statement and no message will be
# printed.
never-returning-functions=sys.exit


[TYPECHECK]

# List of decorators that produce context managers, such as
# contextlib.contextmanager. Add to this list to register other decorators that
# produce valid context managers.
contextmanager-decorators=contextlib.contextmanager

# List of members which are set dynamically and missed by pylint inference
# system, and so shouldn't trigger E1101 when accessed. Python regular
# expressions are accepted.
generated-members=cornice.Service.*,
                  cornice.service.Service.*,
                  str.value
# note: 'str.value' is flagged by Enum members defined with strings

# Tells whether missing members accessed in mixin class should be ignored. A
# mixin class is detected if its name ends with "mixin" (case insensitive).
ignore-mixin-members=yes

# Tells whether to warn about missing members when the owner of the attribute
# is inferred to be None.
ignore-none=yes

# This flag controls whether pylint should warn about no-member and similar
# checks whenever an opaque object is returned when inferring. The inference
# can return multiple potential results while evaluating a Python object, but
# some branches might not be evaluated, which results in partial inference. In
# that case, it might be useful to still emit no-member and other checks for
# the rest of the inferred objects.
ignore-on-opaque-inference=yes

# List of class names for which member attributes should not be checked (useful
# for classes with dynamically set attributes). This supports the use of
# qualified names.
ignored-classes=optparse.Values,thread._local,_thread._local,str,cornice.Service,cornice.service.Service

# List of module names for which member attributes should not be checked
# (useful for modules/projects where namespaces are manipulated during runtime
# and thus existing member attributes cannot be deduced by static analysis). It
# supports qualified module names, as well as Unix pattern matching.
#ignored-modules=

# Show a hint with possible names when a member name was not found. The aspect
# of finding the hint is based on edit distance.
missing-member-hint=yes

# The minimum edit distance a name should have in order to be considered a
# similar match for a missing member name.
missing-member-hint-distance=1

# The total number of similar names that should be taken in consideration when
# showing a hint for a missing member.
missing-member-max-choices=1

# List of decorators that change the signature of a decorated function.
signature-mutators=


[VARIABLES]

# List of additional names supposed to be defined in builtins. Remember that
# you should avoid defining new builtins when possible.
additional-builtins=

# Tells whether unused global variables should be treated as a violation.
allow-global-unused-variables=yes

# List of strings which can identify a callback function by name. A callback
# name must start or end with one of those strings.
callbacks=cb_,
          _cb

# A regular expression matching the name of dummy variables (i.e. expected to
# not be used).
dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_

# Argument names that match this expression will be ignored. Default to name
# with leading underscore.
ignored-argument-names=_.*|^ignored_|^unused_

# Tells whether we should check for unused import in __init__ files.
init-import=yes

# List of qualified module names which can have objects that can redefine
# builtins.
redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io


[SPELLING]

# Limits count of emitted suggestions for spelling mistakes.
max-spelling-suggestions=4

# Spelling dictionary name. Available dictionaries: none. To make it work,
# install the python-enchant package.
spelling-dict=

# List of comma separated words that should not be checked.
spelling-ignore-words=

# A path to a file that contains the private dictionary; one word per line.
spelling-private-dict-file=

# Tells whether to store unknown words to the private dictionary (see the
# --spelling-private-dict-file option) instead of raising a message.
spelling-store-unknown-words=no


[LOGGING]

# Format style used to check logging format string. `old` means using %
# formatting, `new` is for `{}` formatting,and `fstr` is for f-strings.
logging-format-style=old

# Logging modules to check that the string format arguments are in logging
# function parameter format.
logging-modules=logging


[STRING]

# This flag controls whether the implicit-str-concat-in-sequence should
# generate a warning on implicit string concatenation in sequences defined over
# several lines.
check-str-concat-over-line-jumps=no

# plugin:
#   https://pypi.python.org/pypi/pylint-quotes
string-quote=double
triple-quote=double
docstring-quote=double


[FORMAT]

# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
expected-line-ending-format=

# Regexp for a line that is allowed to be longer than the limit.
# - Ignore HTTP(S) URL
# - Ignore RST references (e.g.: long method path)
# Both could be indented (e.g.: within a docstring, class, function, etc.)
ignore-long-lines=^\s*(# )?<?https?://\S+>?$|^\s*\:\w+\:\`\S+\`$

# Number of spaces of indent required inside a hanging or continued line.
indent-after-paren=4

# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
# tab).
indent-string='    '

# Maximum number of characters on a single line.
max-line-length=120

# Maximum number of lines in a module.
max-module-lines=1000

# List of optional constructs for which whitespace checking is disabled. `dict-
# separator` is used to allow tabulation in dicts, etc.: {1  : 1,\n222: 2}.
# `trailing-comma` allows a space between comma and closing bracket: (a, ).
# `empty-line` allows space-only lines.
no-space-check=trailing-comma,
               dict-separator

# Allow the body of a class to be on the same line as the declaration if body
# contains single statement.
single-line-class-stmt=no

# Allow the body of an if to be on the same line as the test if there is no
# else.
single-line-if-stmt=no


[BASIC]

# Naming style matching correct argument names.
argument-naming-style=snake_case

# Regular expression matching correct argument names. Overrides argument-
# naming-style.
#argument-rgx=

# Naming style matching correct attribute names.
attr-naming-style=snake_case

# Regular expression matching correct attribute names. Overrides attr-naming-
# style.
#attr-rgx=

# Bad variable names which should always be refused, separated by a comma.
bad-names=foo,
          bar,
          baz,
          toto,
          tutu,
          tata

# Naming style matching correct class attribute names.
class-attribute-naming-style=any

# Regular expression matching correct class attribute names. Overrides class-
# attribute-naming-style.
#class-attribute-rgx=

# Naming style matching correct class names.
class-naming-style=PascalCase

# Regular expression matching correct class names. Overrides class-naming-style.
# Allow typing definitions that are matched as 'classes' to have slightly more versatile names.
class-rgx=((_{0,2}[A-Z][a-zA-Z0-9]+)|((CWL|PKG|WPS|OAS|JSON|IO|ANY)_[a-zA-Z0-9_]+Types?))$

typealias-rgx=_{0,2}(?!T[A-Z]|Type)((CWL|PKG|WPS|OAS|JSON|IO|ANY)_)?[A-Z]+[a-z0-9]+(?:[A-Z][a-z0-9]+)*$

# Naming style matching correct constant names.
const-naming-style=UPPER_CASE

# Regular expression matching correct constant names. Overrides const-naming-
# style.
#const-rgx=

# Minimum line length for functions/classes that require docstrings, shorter
# ones are exempt.
docstring-min-length=-1

# Naming style matching correct function names.
function-naming-style=snake_case

# Regular expression matching correct function names. Overrides function-
# naming-style.
#function-rgx=

# Good variable names which should always be accepted, separated by a comma.
good-names=i,j,k,v,kv,ex,x,y,z,f,h,db,kw,dt,q,ns,id,s3,to,_

# Include a hint for the correct naming format with invalid-name.
include-naming-hint=yes

# Naming style matching correct inline iteration names.
inlinevar-naming-style=any

# Regular expression matching correct inline iteration names. Overrides
# inlinevar-naming-style.
#inlinevar-rgx=

# Naming style matching correct method names.
method-naming-style=snake_case

# Regular expression matching correct method names. Overrides method-naming-
# style.
#method-rgx=

# Naming style matching correct module names.
module-naming-style=snake_case

# Regular expression matching correct module names. Overrides module-naming-
# style.
#module-rgx=

# Colon-delimited sets of names that determine each other's naming style when
# the name regexes allow several styles.
name-group=

# Regular expression which should only match function or class names that do
# not require a docstring.
no-docstring-rgx=^_

# List of decorators that produce properties, such as abc.abstractproperty. Add
# to this list to register other decorators that produce valid properties.
# These decorators are taken in consideration only for invalid-name.
property-classes=abc.abstractproperty

# Naming style matching correct variable names.
variable-naming-style=snake_case

# Regular expression matching correct variable names. Overrides variable-
# naming-style.
#variable-rgx=


[MISCELLANEOUS]

# List of note tags to take in consideration, separated by a comma.
notes=HACK
## FIXME,TODO,


[SIMILARITIES]

# Ignore comments when computing similarities.
ignore-comments=yes

# Ignore docstrings when computing similarities.
ignore-docstrings=yes

# Ignore imports when computing similarities.
ignore-imports=yes

# Minimum lines number of a similarity.
min-similarity-lines=10


[CLASSES]

# List of method names used to declare (i.e. assign) instance attributes.
defining-attr-methods=__init__,
                      __new__,
                      setUp,
                      __post_init__

# List of member names, which should be excluded from the protected access
# warning.
exclude-protected=_asdict,
                  _fields,
                  _replace,
                  _source,
                  _make

# List of valid names for the first argument in a class method.
valid-classmethod-first-arg=cls

# List of valid names for the first argument in a metaclass class method.
valid-metaclass-classmethod-first-arg=cls


[IMPORTS]

# List of modules that can be imported at any level, not just the top level
# one.
allow-any-import-level=

# Allow wildcard imports from modules that define __all__.
allow-wildcard-with-all=no

# Analyse import fallback blocks. This can be used to support both Python 2 and
# 3 compatible code, which means that the block might have code that exists
# only in one or another interpreter, leading to false positives when analysed.
analyse-fallback-blocks=no

# Deprecated modules which should not be used, separated by a comma.
deprecated-modules=optparse,tkinter.tix

# Create a graph of external dependencies in the given file (report RP0402 must
# not be disabled).
ext-import-graph=

# Create a graph of every (i.e. internal and external) dependencies in the
# given file (report RP0402 must not be disabled).
import-graph=

# Create a graph of internal dependencies in the given file (report RP0402 must
# not be disabled).
int-import-graph=

# Force import order to recognize a module as part of the standard
# compatibility libraries.
known-standard-library=posixpath,typing,typing_extensions

# Force import order to recognize a module as part of a third party library.
known-third-party=enchant,cornice_swagger,cwltool,cwt,docker

# Couples of modules and preferred modules, separated by a comma.
preferred-modules=


[DESIGN]

# Maximum number of arguments for function / method.
max-args=20

# Maximum number of attributes for a class (see R0902).
max-attributes=20

# Maximum number of boolean expressions in an if statement (see R0916).
max-bool-expr=20

# Maximum number of branch for function / method body.
max-branches=20

# Maximum number of locals for function / method body.
max-locals=20

# Maximum number of parents for a class (see R0901).
max-parents=10

# Maximum number of public methods for a class (see R0904).
max-public-methods=20

# Maximum number of return / yield for function / method body.
max-returns=20

# Maximum number of statements in function / method body.
max-statements=100

# Minimum number of public methods for a class (see R0903).
min-public-methods=0


[EXCEPTIONS]

# Exceptions that will emit a warning when being caught. Defaults to
# "BaseException, Exception".
overgeneral-exceptions=builtins.BaseException
```

### Command used

```shell
pylint \
  --rcfile="$(APP_ROOT)/.pylintrc" \
  "$(APP_ROOT)/weaver" "$(APP_ROOT)/tests"
```

With https://github.com/crim-ca/weaver/blob/refs/heads/dependabot/pip/pylint-gte-4.0.5-and-lt-5/.pylintrc

### Pylint output

```python
************* Module weaver.xml_util
weaver/xml_util.py:71:0: C0103: Variable name "HTML" doesn't conform to snake_case naming style ('([^\\W\\dA-Z][^\\WA-Z]*|_[^\\WA-Z]*|__[^\\WA-Z\\d_][^\\WA-Z]+__)$' pattern) (invalid-name)
```

### Expected behavior

It doesn't look like it should raise a "variable name" error, since it is all caps (constant at module level).

### Pylint version

```shell
pylint 4.0.5
astroid 4.0.4
Python 3.12.12 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 20:16:04) [GCC 11.2.0]
```

### Additional dependencies

```python

```

**Comment by Socialpranker** [comment]
I tried to reproduce this to work on a fix, but couldn't:

- Installed the exact reported versions (`pylint==4.0.5`, `astroid==4.0.4`, Python 3.12.13) against the exact file from the linked commit (`weaver/xml_util.py` at `8ec0359`), both in isolation and inside a `weaver` package so the module context matches.
- Also tried current `main` (pylint 4.1.0.dev0, astroid 4.2.0b4) and Python 3.14.

In every case `HTML = lxml_etree.HTML` scores 10.00/10 with `invalid-name` enabled — no warning is raised.

My guess is this is sensitive to how `lxml` was built/installed in the reporter's environment (the traceback mentions `packaged by Anaconda, Inc.`), since `invalid-name`'s constant-vs-variable classification for a module-level assignment depends on `astroid` successfully inferring the assigned value's type (`visit_assignname` in `checker.py`, around the `inferred_assign_type` handling) — and C-extension inference for something like `lxml.etree.HTML` can vary depending on how the extension module was compiled/introspected.

Not confident enough to guess at a fix without being able to reproduce it. If anyone can share a minimal `conda`-based repro (or confirm the OS/lxml build), that would help narrow it down.

**Comment by fmigneault** [comment]
Is the "packaged by Anaconda, Inc." on your end caused by install via `environment.yml` or similar in conda env? That is not my typical installation (and probably not most user's setup since Anaconda imposes commercial license restrictions while `nodefaults`/`conda-forge` channels are open). 

I do use conda locally for development and managing envs, but all packages are installed via `pip` requirements (https://github.com/crim-ca/weaver/blob/master/requirements.txt) where `lxml` can be found. 

In the CI (where the pylint logs are obtained as shown below), it is all running in the Python setup by GitHub without any conda involved.

https://github.com/crim-ca/weaver/pull/956
https://github.com/crim-ca/weaver/actions/runs/25706245193/job/75476879490?pr=956
```
Running linting code style checks...
************* Module weaver.xml_util
weaver/xml_util.py:71:0: C0103: Variable name "HTML" doesn't conform to snake_case naming style ('([^\\W\\dA-Z][^\\WA-Z]*|_[^\\WA-Z]*|__[^\\WA-Z\\d_][^\\WA-Z]+__)$' pattern) (invalid-name)
************* Module weaver.datatype
weaver/datatype.py:138:4: C0103: Constant name "_Authentication" doesn't conform to UPPER_CASE naming style ('([^\\W\\da-z][^\\Wa-z]*|__.*__)$' pattern) (invalid-name)
weaver/datatype.py:139:4: C0103: Constant name "_DockerAuthentication" doesn't conform to UPPER_CASE naming style ('([^\\W\\da-z][^\\Wa-z]*|__.*__)$' pattern) (invalid-name)
************* Module weaver.wps_restapi.jobs.utils
weaver/wps_restapi/jobs/utils.py:1673:4: E1206: Not enough arguments for logging format string (logging-too-few-args)

------------------------------------
Your code has been rated at 10.00/10
```
